src/vs/workbench/contrib/notebook/browser/contrib/saveParticipants/saveParticipants.ts

```diff
@@ -86,7 +91,217 @@ class FormatOnSaveParticipant implements IStoredFileWorkingCopySaveParticipant {
+				if (viewCell.model.textModel) {
+					cannotTouchLineNumber = Math.max(cannotTouchLineNumber, sel.selectionStartLineNumber);
+				}
+			}
+		}
+
+
+		try {
+			const allCellEdits = await Promise.all(notebook.cells.map(async (cell) => {
+				if (cell.cellKind !== 2) {
+					return;
+				}
+
+				const ref = await this.textModelService.createModelReference(cell.uri);
+				disposable.add(ref);
+				const model = ref.object.textEditorModel;
+
+				const lastNonEmptyLine = this.findLastNonEmptyLine(model);
+				const deleteFromLineNumber = Math.max(lastNonEmptyLine + 1, cannotTouchLineNumber + 1);
+				const deletionRange = model.validateRange(new Range(deleteFromLineNumber, 1, model.getLineCount(), model.getLineMaxColumn(model.getLineCount())));
+
+				if (deletionRange.isEmpty()) {
+					return;
+				}
+
+				// create the edit to delete all lines in deletionRange
+				return new ResourceTextEdit(model.uri, { range: deletionRange, text: '' }, model.getVersionId());
+			}));
+
+			const filteredEdits = allCellEdits.flat().filter(edit => edit !== undefined) as ResourceEdit[];
+			await this.bulkEditService.apply(filteredEdits, { label: localize('trimNotebookNewlines', "Trim Final New Lines"), code: 'undoredo.trimFinalNewLines' });
+
+		} finally {
+			progress.report({ increment: 100 });
+			disposable.dispose();
+		}
+	}
+}
+
+class FinalNewLineParticipant implements IStoredFileWorkingCopySaveParticipant {
+
+	constructor(
+		@IConfigurationService private readonly configurationService: IConfigurationService,
+		// @IEditorService private readonly editorService: IEditorService,
+		@ITextModelService private readonly textModelService: ITextModelService,
+		@IBulkEditService private readonly bulkEditService: IBulkEditService,
+	) { }
+
+	async participate(workingCopy: IStoredFileWorkingCopy<IStoredFileWorkingCopyModel>, context: { reason: SaveReason }, progress: IProgress<IProgressStep>, _token: CancellationToken): Promise<void> {
+		if (this.configurationService.getValue('files.insertFinalNewline')) {
+			this.doInsertFinalNewLine(workingCopy, context, progress);
+		}
+	}
+
+	private async doInsertFinalNewLine(workingCopy: IStoredFileWorkingCopy<IStoredFileWorkingCopyModel>, context: { reason: SaveReason }, progress: IProgress<IProgressStep>): Promise<void> {
+		if (!workingCopy.model || !(workingCopy.model instanceof NotebookFileWorkingCopyModel)) {
+			return;
+		}
+
+		const disposable = new DisposableStore();
+		const notebook = workingCopy.model.notebookModel;
+
+		try {
+			const allCellEdits = await Promise.all(notebook.cells.map(async (cell) => {
+				if (cell.cellKind !== 2) {
+					return;
+				}
+
+				const ref = await this.textModelService.createModelReference(cell.uri);
+				disposable.add(ref);
+				const model = ref.object.textEditorModel;
+
+				const lineCount = model.getLineCount();
+				const lastLine = model.getLineContent(lineCount);
+				const lastLineIsEmptyOrWhitespace = strings.lastNonWhitespaceIndex(lastLine) === -1;
+
```

Comment:
Same here, let's use `textBuffer` so we don't have to create a temporary `textModel`. We can check if the last line is empty by comparing `getLineLastNonWhitespaceColumn` and `getLineMaxColumn`